core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala

```diff
@@ -14,34 +14,35 @@ import scodec.bits.ByteVector
 sealed abstract class ScriptWitness extends NetworkElement {
 
   /** The byte vectors that are placed on to the stack when evaluating a witness program */
-  def stack: Seq[ByteVector]
+  val stack: Seq[ByteVector]
 
-  override def bytes = RawScriptWitnessParser.write(this)
 }
 
 case object EmptyScriptWitness extends ScriptWitness {
-  override def stack = Nil
+  override val stack: Seq[ByteVector] = Vector.empty
 
-  override def bytes = ByteVector.low(1)
+  override val bytes: ByteVector = ByteVector.low(1)
 }
 
-sealed abstract class ScriptWitnessV0 extends ScriptWitness
+sealed abstract class ScriptWitnessV0 extends ScriptWitness {
+  override val bytes: ByteVector = RawScriptWitnessParser.write(this)
+}
 
 /**
  * Represents a [[org.bitcoins.core.protocol.script.ScriptWitness]] that is needed to spend a
  * [[P2WPKHWitnessV0]] scriptPubKey
  * [[https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#p2wpkh-nested-in-bip16-p2sh]]
  * Format: <pubKey> <signature>
  */
-sealed abstract class P2WPKHWitnessV0 extends ScriptWitness {
+sealed abstract class P2WPKHWitnessV0 extends ScriptWitnessV0 {
   def pubKey: ECPublicKey = ECPublicKey(stack.head)
 
   def signature: ECDigitalSignature = stack(1) match {
     case ByteVector.empty => EmptyDigitalSignature
     case bytes: ByteVector => ECDigitalSignature(bytes)
   }
 
-  override def toString = "P2WPKHWitnessV0(" + stack.map(BitcoinSUtil.encodeHex(_)).toString + ")"
+  override def toString = s"P2WPKHWitnessV0(${stack.map(BitcoinSUtil.encodeHex(_)).toString})"
```

Comment:
Is this just for readability, or are there performance issues here?

Reply:
technically there are performance issues on multiple fronts

1. concatenating strings are generally expensive on the jvm [source](https://stackoverflow.com/questions/10078912/best-practices-performance-mixing-stringbuilder-append-with-string-concat)

2. i should really be caching the val 
```
stack.map(BitcoinSUtil.encodeHex(_)).toString
```